### DISCUSSION TITLE
Brainstorming for a UnityAtoms v5 / Rewrite / Successor

### DISCUSSION THREAD

**Discussion by soraphis** [unity-atoms/unity-atoms#397] [General]
DISCLAIMER 1: _the following is full of personal opinion_
DISCLAIMER 2: _this is just a bit of brainstorming, on things that work nice in UnityAtoms, and things that don't_

### Good things in UnityAtoms v4.x
- There are a lot of good things in atoms, but I'm too lazy to list them all :D

### Bad things in UnityAtoms v4.x
- Inheritance Architecture. This was chosen because it's easier for the serialization system of unity.
- AtomActions are to cumbersome to work with. The Idea is great, the execution is meh.
- Addressables and SceneLoading messing with the lifecycle of ScriptableObjects (often different in a build than in editor)
- A lot of ScripableObjects → cluttering the project folder and makes it hard to scope variables (what is global state, and which one is only used in scene 1 as a temporary value?)
- AtomList / AtomCollection, they are named too-similar to AtomValueList but they are
- Instancers, runtime creation of ScriptableObjects have a too high performance impact, that is easily overlooked in the beginning of an project.
- Events. In general they are nice, but they mix two different concepts: Events can be an "EventBus", there is no state, there are just messages send on that line, but they can also be the ChangedEvent on a variable, and then they suddenly have an "initial-state" (at least conceptually).

### Things I'd like to do different in the future

- ScriptableObjects should be runtime-immutable. Runtime-Values should be stored in a "VariableManager" (Self-Instantiating-Singleton, don't destroy on load). This way...
    - ... a) ending playmode automatically loses the runtime values, no extra reset necessary, works easily with disabled domain reload
    - ... b) references to Variables are kept and SceneLoading (in a build) does not free the resources and resets their state 
- AtomActions could make use of `[SerializedReference]` (which has it's own drawbacks, sadly)
    - this removes the need to create an asset instance (though it creates a class-instance at runtime/deserialization, but that should be okay (compared to the overhead a SO has and considering the lifetime is as long as the containing object))
    - this allows different actions to have different fields that can be configured on the component itself.
- AtomActions need a way to pass arguments to them dynamically. 
- Scoping/Grouping. Make it easy to manage which variable are globally scoped, scene scoped, gameobject/prefab scoped.
- No replay buffer (https://github.com/unity-atoms/unity-atoms/issues/194#issuecomment-699643908)

**Comment by jackboz** [comment]
Additional wishes on atom actions.

If I want to call an atom action in OnSomethingHook I have to do the following:
- add a OnSomethingHook component.
- add a Void Event Reference Listener.
- create a VoidEvent and assign it to the listener and the Hook
- create a custom AtomAction that does your job
- create an instance of that AtomAction and assign it on the listener to the ActionResponses

But it seems that one can avoid steps on creating an additional event and an corresponding listener. We can modify an OnSomethingHook that it can call an action directly.

**Reply by soraphis** [reply]
I'm dangling a bit in small prototype project right now, there I actually went with an `[SerializedReference]` + [`[SubclassSelector]`](https://github.com/mackysoft/Unity-SerializeReferenceExtensions) combo.

so i have an abstract base of SomeAction -> and have a child class for a UntiyAction, or other specific actions. 
using these feels less cluttered and more flexible imho :)

But yeah, On<>Hook should probably support AtomActions .. could be a 4.7 feature

**Comment by builder-main** [comment]
For your main point, about lifcycle of the scriptables. It sounds you would need to enforce instantiation, however it seems a bit cumbersome ?
Couldn't it be a project architecture prerogative ? You could define "Gameplay/Runtime" scriptables which are here on the purpose of databinding stores and config scriptable objects which are static source of data.

**Reply by soraphis** [reply]
> For your main point, about lifcycle of the scriptables. It sounds you would need to enforce instantiation, however it seems a bit cumbersome ?

I'm not sure I get what you're saying. 

Right now there is not much control we can achive from UnityAtoms about asset lifecycles. that is highly project dependant and is vastly different when using Addressables and some (wrong or inefficient) Addressable configurations might lead to ScriptableObjects being loaded multiple times. these have the same name and intial value but could have two different runtime values and being loaded at the same time.
It's extremely difficult to debug and figure out, happens ONLY in a build, and UnityAtoms can't really do anything about it.

Core of this issue is, that the actual loaded asset stores the value. 

This also has the issue of "cleaning" in Edit mode, since Assets keep their value when doing runtime modifications. so we need extra lifecycle events for cleaning when leaving the edit mode.

In a better world Atoms would be immutable at runtime. so they are just used as keys (and maybe initial values and things like that), the value is stored in some runtime only container.
